### PR TITLE
EOS-26647 : Multipart object replication

### PR DESCRIPTION
**Pull request by priyanka25081999**
# Problem Statement
- Updated the complete multipart upload class to get the request state
- Created a new class of multipart object replicator to add functionality of get range read and multipart PUT.
- Implemented the multipart transfer test and added a new config file
- Added few future references
Signed-off-by: Priyanka Salunke <priyanka.s.salunke@seagate.com>

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by priyanka25081999** [comment]
Jenkins build is passed - http://eos-jenkins.colo.seagate.com/job/Multisite/job/Run-Tests/20/

File: s3/replication/replicator/src/s3replicator/transfer_initiator.py [hunk 1]
```diff
@@ -89,37 +91,94 @@ async def start(job, app):
...
-                "all_events", TranferEventHandler(app))
+            # calculate single part length
+            part_len = object_size / int(total_parts)
 
-            job.set_replicator(object_tag_replicator)
+            # Append the length of each part into the list
+            # Pass the list to multipart object replicator class
+            for part_no in range(1, int(total_parts) + 1):
+                _logger.debug("Part No : {}".format(part_no))
+                part_length.append(int(part_len))
```
Comment (huanghua78) [inline s3/replication/replicator/src/s3replicator/transfer_initiator.py:122]: suppose we have an object, 100MB.
And every part is 15MB.
So, we will have 7 parts, and their sizes are:
part 1: 15MB
part 2: 15MB
part 3: 15MB
part 4: 15MB
part 5: 15MB
part 6: 15MB
part 7: 10MB

We are not dividing the objects into parts with equal sizes.
All parts except the last one have same sizes.
Only the last part could be less than standard size.

Comment (priyanka25081999) [reply s3/replication/replicator/src/s3replicator/transfer_initiator.py:122]: Yes, this is correct.
Currently, Replicator is dividing the object by the part number as we are not getting specific part details. we are assuming that all parts are of fixed size. 
I have added few future references in the code. So, once we get the updated head object and get object api, we will modify the code and the replicator will only get specific parts from the source and put them in the destination.

**Review by huanghua78 — APPROVED** [review]

**Review by mbcortx — APPROVED** [review]

**Comment by johnbent** [comment]
I was going to merge this but github won't let me because it is marked as draft. Please remove draft status if you want me to merge this.

**Comment by priyanka25081999** [comment]
@johnbent I have removed the PR from draft. Please merge it.

**Pull request merged**